### PR TITLE
103 mapbinder

### LINKED ISSUE #103: EntityMapBinder: whiteList and blackList
need to be able to pass these in. 

should also lean on the field added to the domain entities by default `$defaultDatabindingWhiteList`
DefaultASTDatabindingHelper.DEFAULT_DATABINDING_WHITELIST and its use can be seen in Grails source getBindingIncludeList in DataBindingUtils along with how it caches it in CLASS_TO_BINDING_INCLUDE_LIST.

### PR DESCRIPTION
**Pull request by snimavat**
closes #103 #102

### CODE REVIEW

File: plugin/src/main/groovy/gorm/tools/databinding/EntityMapBinder.groovy [hunk 1]
```diff
@@ -2,38 +2,63 @@ package gorm.tools.databinding
...
+import java.util.concurrent.ConcurrentHashMap
+
 /**
  * Faster data binder for PersistentEntity.persistentProperties. Uses the persistentProperties to assign values from the Map
  * Explicitly checks and converts most common property types eg (numbers and dates). Otherwise fallbacks to value converters.
  *
  */
 @CompileStatic
 class EntityMapBinder implements MapBinder {
+    private static final Map<Class, List> CLASS_TO_BINDING_INCLUDE_LIST = new ConcurrentHashMap<Class, List>()
```
Comment (basejump) [inline plugin/src/main/groovy/gorm/tools/databinding/EntityMapBinder.groovy]: rename to be consitent

File: plugin/src/main/groovy/gorm/tools/databinding/EntityMapBinder.groovy [hunk 2]
```diff
@@ -84,4 +113,45 @@ class EntityMapBinder implements MapBinder {
...
+                }
+            }
+        } catch (Exception e) {
+        }
+        return includeList
+    }
+
+    protected preprocessValue(propertyValue) {
+        if(propertyValue instanceof CharSequence) {
+            String stringValue = propertyValue.toString()
```
Comment (basejump) [inline plugin/src/main/groovy/gorm/tools/databinding/EntityMapBinder.groovy]: toString or cast?

File: plugin/src/main/groovy/gorm/tools/databinding/MapBinder.groovy [hunk 3]
```diff
@@ -1,18 +1,17 @@
...
  */
 interface MapBinder {
 
     void bind(Object target, Map<String, Object> source, BindAction bindAction)
-
     void bind(Object target, Map<String, Object> source)
 
-    void bindCreate(Object target, Map<String, Object> source)
+    void bind(Object target, Map<String, Object> source, BindAction bindAction, List<String> whiteList, List<String> blackList)
+    void bind(Object target, Map<String, Object> source, List<String> whiteList, List<String> blackList)
```
Comment (basejump) [inline plugin/src/main/groovy/gorm/tools/databinding/MapBinder.groovy]: I think we should do this with an optional args map so we can do something more like `mapBinder.bind(user, p, bindAction, include: bindInclude)` or `mapBinder.bind(user, p, exclude:['foo','bar'])`

File: plugin/src/main/groovy/gorm/tools/hibernate/criteria/GormHibernateCriteriaBuilder.groovy [hunk 4]
```diff
@@ -28,12 +28,8 @@ import org.hibernate.SessionFactory
  * order('invoice.customer.name')
  *
  * instead of
- * invoice {
- *    customer {
- *       order(name)
- *    }
- * }
- *
+ * invoice {*    customer {*       order(name)
```
Comment (basejump) [inline plugin/src/main/groovy/gorm/tools/hibernate/criteria/GormHibernateCriteriaBuilder.groovy]: format scrambled

File: plugin/src/main/groovy/gorm/tools/databinding/EntityMapBinder.groovy (hunk 2 above)
Comment (snimavat) [reply plugin/src/main/groovy/gorm/tools/databinding/EntityMapBinder.groovy]: Just used from the grails data binder.. can just check for string and cast

**Comment by basejump** [comment]
@snimavat performance degrades with these changes.
```
java -server -jar -DrunDataBindingBenchmark=true build/libs/benchmarks.war
 - setters, copy and fast bind on simple no associations
1.037s setPropsFastIterate CityFatSimple | 111690 rows

 - setters or property copy on associations with 20 fields
3.428s useStaticSettersInDomain CityFat | 111690 rows
5.911s useFastBinder CityFat | 111690 rows
```

```
gradle assemble -DauditTrailEnabled=false --offline; java -server -jar -DauditTrailEnabled=false -DloadIterations=3 -Dgpars.poolsize=1 build/libs/benchmarks.war
### Gpars - Assign Properties, no grails databinding
  - Grails Basic Baseline to measure against
6.384s for GparsBaselineBenchmark<CityBaseline> [ dataBinder: fast ]
6.484s for GparsBaselineBenchmark<City> [ dataBinder: fast ]
  - Events disabled
6.817s for GparsRepoBenchmark<City> [ dataBinder: fast ]
  - Events enabled
7.889s for GparsRepoBenchmark<City> [ dataBinder: fast ]
```

On Master its
```
 - setters or property copy on associations with 20 fields
3.436s useStaticSettersInDomain CityFat | 111690 rows
4.569s useFastBinder CityFat | 111690 rows
.......
### Gpars - Assign Properties, no grails databinding
  - Grails Basic Baseline to measure against
5.834s for GparsBaselineBenchmark<CityBaseline> [ dataBinder: fast ]
5.893s for GparsBaselineBenchmark<City> [ dataBinder: fast ]
```

**Pull request closed without merging**